packages/debug/src/browser/abstract-launch-preference-provider.ts

```diff
@@ -0,0 +1,157 @@
+/********************************************************************************
+ * Copyright (C) 2019 Red Hat, Inc. and others.
+ *
+ * This program and the accompanying materials are made available under the
+ * terms of the Eclipse Public License v. 2.0 which is available at
+ * http://www.eclipse.org/legal/epl-2.0.
+ *
+ * This Source Code may also be made available under the following Secondary
+ * Licenses when the conditions for such availability set forth in the Eclipse
+ * Public License v. 2.0 are satisfied: GNU General Public License, version 2
+ * with the GNU Classpath Exception which is available at
+ * https://www.gnu.org/software/classpath/license.html.
+ *
+ * SPDX-License-Identifier: EPL-2.0 OR GPL-2.0 WITH Classpath-exception-2.0
+ ********************************************************************************/
+
+import { injectable, postConstruct } from 'inversify';
+import { PreferenceScope, PreferenceProvider } from '@theia/core/lib/browser';
+import { Emitter, Event } from '@theia/core';
+import { Deferred } from '@theia/core/lib/common/promise-util';
+import { DisposableCollection } from '@theia/core';
+import { Disposable } from '@theia/core';
+
+export interface GlobalLaunchConfig {
+    version: string;
+    compounds?: LaunchCompound[];
+    configurations: LaunchConfig[];
+}
+
+export namespace GlobalLaunchConfig {
+    /* tslint:disable-next-line:no-any */
+    export function is(data: any): data is GlobalLaunchConfig {
+        return !data || (!!data.version && (!data.compounds || Array.isArray(data.compounds)) && Array.isArray(data.configurations));
+    }
+}
+
+export interface LaunchConfig {
+    type: string;
+    request: string;
+    name: string;
+
+    /* tslint:disable-next-line:no-any */
+    [field: string]: any;
+}
+
+export interface LaunchCompound {
+    name: string;
+    configurations: (string | { name: string, folder: string })[];
+}
+
+export const LaunchPreferenceProvider = Symbol('LaunchConfigurationProvider');
+export interface LaunchPreferenceProvider {
+
+    readonly onDidLaunchChanged: Event<void>;
+
+    ready: Promise<void>;
+
+    getConfigurationNames(withCompounds: boolean, resourceUri?: string): string[];
+
+}
+
+export const FolderLaunchProviderOptions = Symbol('FolderLaunchProviderOptions');
+export interface FolderLaunchProviderOptions {
+    folderUri: string;
+}
+
+export const LaunchProviderProvider = Symbol('LaunchProviderProvider');
+export type LaunchProviderProvider = (scope: PreferenceScope) => LaunchPreferenceProvider;
+
+@injectable()
+export abstract class AbstractLaunchPreferenceProvider implements LaunchPreferenceProvider, Disposable {
```

Comment:
I fail to understand why stuff from launch configuration should end up in the preferences. Can you explain the motivation and concept here, please?

Reply:
Also, how does it relate to fixing the original bug?

Reply:
the motivation and concept was explained here, i guess: https://github.com/theia-ide/theia/pull/4743/files#r270355020

Reply:
In accordance to VSCode API, it should be possible to retrieve configuration from `launch.json`  as preferences. 
[VS Code API](https://code.visualstudio.com/api/references/vscode-api#WorkspaceConfiguration)

Also, User Settings could contains a global debug launch configuration as alternative to `launch.json`.
[User and Workspace settings](https://code.visualstudio.com/docs/getstarted/settings#_default-settings)

Reply:
> Also, how does it relate to fixing the original bug?

It provides ability for a plugin to get launch configurations using Preferences API.